tree/ntuple/v7/src/RNTupleWriteOptions.cxx

```diff
@@ -48,18 +52,38 @@ std::unique_ptr<ROOT::Experimental::RNTupleWriteOptions> ROOT::Experimental::RNT
 
 void ROOT::Experimental::RNTupleWriteOptions::SetApproxZippedClusterSize(std::size_t val)
 {
-   EnsureValidTunables(val, fMaxUnzippedClusterSize, fApproxUnzippedPageSize);
+   EnsureValidTunables(val, fMaxUnzippedClusterSize, fInitialNElementsPerPage, fMaxUnzippedPageSize);
    fApproxZippedClusterSize = val;
 }
 
 void ROOT::Experimental::RNTupleWriteOptions::SetMaxUnzippedClusterSize(std::size_t val)
 {
-   EnsureValidTunables(fApproxZippedClusterSize, val, fApproxUnzippedPageSize);
+   EnsureValidTunables(fApproxZippedClusterSize, val, fInitialNElementsPerPage, fMaxUnzippedPageSize);
    fMaxUnzippedClusterSize = val;
 }
 
 void ROOT::Experimental::RNTupleWriteOptions::SetApproxUnzippedPageSize(std::size_t val)
 {
-   EnsureValidTunables(fApproxZippedClusterSize, fMaxUnzippedClusterSize, val);
+   // TODO(jblomer): remove option
    fApproxUnzippedPageSize = val;
 }
+
+void ROOT::Experimental::RNTupleWriteOptions::SetInitialNElementsPerPage(std::size_t val)
+{
+   EnsureValidTunables(fApproxZippedClusterSize, fMaxUnzippedClusterSize, val, fMaxUnzippedPageSize);
+   fInitialNElementsPerPage = val;
+}
+
+void ROOT::Experimental::RNTupleWriteOptions::SetMaxUnzippedPageSize(std::size_t val)
+{
+   EnsureValidTunables(fApproxZippedClusterSize, fMaxUnzippedClusterSize, fInitialNElementsPerPage, val);
+   fMaxUnzippedPageSize = val;
+}
+
+std::size_t ROOT::Experimental::RNTupleWriteOptions::GetPageBufferBudget() const
+{
+   if (fPageBufferBudget != 0)
+      return fPageBufferBudget;
+
+   return GetApproxZippedClusterSize() + (GetCompression() != 0) * GetApproxZippedClusterSize();
```

Comment:
I think we should consider adding a factor 1.5x to 2x here to account for allocated, but unused buffer space during expansion. I believe this should greatly reduce the number of pages with the simplified flushing algorithm only considering the current column.

Reply:
If I add a factor two, the number of pages is indeed much lower (a little lower even than the "adaptive-50" column in the table). However, I think that's beside the point. In the case of an increased budget, we simply never exhaust it (tested). The question of the flushing algorithm is relevant for the case when we actually do exhaust the page buffer budget.

I added a branch with the [experimental changes](https://github.com/jblomer/root/tree/ntuple-adaptive-page-sizes-v2-use-exp) used in the test.

Reply:
Right, it was my understanding that we want to do this because we can precisely predict which memory budget we need to achieve this effect in practice. However you're probably right that it's better to consume less memory if we can get the same result with a bit more complicated code...